### ISSUE TITLE
Add overload decorator

### ISSUE THREAD

**Issue by said-hadjout** [pyccel/pyccel#1767]
As a user I want to define multiple functions with different bodies and signatures but with the same name, e.g.
```python
@overload
def f(x:'int[:]'):
    return x.sum()

@overload
def f(x:'float[:]'):
    return x.max()
``

**Comment by EmilyBourne** [comment]
I am closing this issue as a duplicate of #1334 .
There have been a few discussions but there does not seem to be a good consensus about the best way to achieve this yet.

**Comment by said-hadjout** [comment]
I think we already have the mechanism, in C there is no concept of Interface, so it should be straightforward there, in Fortran it will be treated like C

**Comment by EmilyBourne** [comment]
Yes, the problem is how to write this in Python. There have been suggestions of using `@overload` from the typing module, or `@singledispatch` from [`functools`](https://docs.python.org/3/library/functools.html#functools.singledispatch) or `@dispatch` from [multipledispatch](https://pypi.org/project/multipledispatch/). My preference would be either `@overload` or `@singledispatch`. After #1728 `@overload` should be trivial to implement so I agree that this is the best solution. However as the discussion already exists, and was already created as an issue after discussion and then turned into a discussion again I think it is important to discuss this there first to ensure a consensus is reached. This will then allow us to open an issue following the "Feature Request" template which asks for a relevant discussion to be referenced
